File: swift/1679_MaxNumberOfK-SumPairs.swift
class Solution {
    func maxOperations(_ nums: [Int], _ k: Int) -> Int {
        var sortedNums = nums.sorted()
        var left = 0, right = sortedNums.count - 1
        var count = 0
        while left < right {
            var total = sortedNums[left] + sortedNums[right]
            if total == k {
                left += 1
                right -= 1
                count += 1
            } else if total > k {
                right -= 1
            } else {
                left += 1
            }
        }
        return count
    }
}

class Solution {
    func maxOperations(_ nums: [Int], _ k: Int) -> Int {
        var numCounts = [Int:Int]()
        var count = 0
        for n in nums {
            let complement = k - n
            if let currentCount = numCounts[complement], currentCount > 0 {
                numCounts[complement]! -= 1
                count += 1
            } else {
                numCounts[n] = numCounts[n] == nil ? 1 : numCounts[n]! + 1
            }
        }
        return count
    }
}

File: java/1679_MaxNumberOfK-SumPairs.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        Arrays.sort(nums);
        int left = 0, right = nums.length - 1, count = 0;
        while (left < right) {
            int total = nums[left] + nums[right];
            if (total == k) {
                left++;
                right--;
                count++;
            } else if (total < k) {
                left++;
            } else {
                right--;
            }
        }
        return count;
    }
}

class Solution {
    public int maxOperations(int[] nums, int k) {
        HashMap<Integer, Integer> numCounts = new HashMap<Integer, Integer>();
        int count = 0;
        for (int n : nums) {
            int complement = k - n;
            if (numCounts.getOrDefault(complement, 0) > 0) {
                numCounts.put(complement, numCounts.get(complement) - 1);
                count++;
            } else {
                numCounts.put(n, numCounts.getOrDefault(n, 0) + 1);
            }
        }
        return count;
    }
}
